### PR TITLE
The approuter controller bugs fix

### PR DESCRIPTION
**Pull request by wchao1241**
fix issue [#13833](https://github.com/rancher/rancher/issues/13833)

1. Fix the approuter controller crash
2. Make sure each ingress hostname is valid

### CODE REVIEW

File: pkg/controllers/user/approuter/ingress_controller.go [hunk 1]
```diff
@@ -44,6 +43,10 @@ func isGeneratedDomain(obj *extensionsv1beta1.Ingress, host, domain string) bool
 }
 
 func (c *Controller) sync(key string, obj *extensionsv1beta1.Ingress) error {
+	if obj == nil {
```
Comment (alena1108) [inline pkg/controllers/user/approuter/ingress_controller.go]: should we check for obj.DeletionTimestamp != nil too? both can indicate the removed object

Comment (alena1108) [reply pkg/controllers/user/approuter/ingress_controller.go]: also I'd think that underlying code should cleanup dns record for the ingress that had been removed, right? even if the sync method returns

Comment (wchao1241) [reply pkg/controllers/user/approuter/ingress_controller.go]: The rdns-server(coredns) can support wildcard records, so only one record will be stored for a k8s cluster,all the ingress resources in a cluster use the dns record.The record will be automatically removed only after the k8s cluster is destroyed.

Comment (wchao1241) [reply pkg/controllers/user/approuter/ingress_controller.go]: @alena1108 yes,It's really necessary to check `obj.DeletionTimestamp != nil` and I have added it.

Comment (alena1108) [reply pkg/controllers/user/approuter/ingress_controller.go]: Nit: can we do it in one line? if `obj == nil ||`. Also `obj.DeletionTimestamp` - you can access directly, no need to reference via objectMeta.

**Comment by wchao1241** [comment]
Add commit (id: 6a88cae1c7e5c375297a0404ce955ffd78a6bbcc)
fix the following issues:
1.fix issue [#13848](https://github.com/rancher/rancher/issues/13848)
2.it seems like token is not required to get a domain.
3.every call to ApplyDomain does a GET to the public service.  This seems excesive as ApplyDomain is called for every ingress change.  Also it seems like a domain is created even if setting is not "lb.rancher.cloud"

**Pull request merged**